common/flatpak-dir.c

```diff
@@ -2393,6 +2393,281 @@ flatpak_dir_find_latest_rev (FlatpakDir               *self,
+
+      latest_rev = find_latest_p2p_result (results, &data->collection_ref);
+      if (g_strcmp0 (latest_rev, data->local_commit) == 0)
+        {
+          g_autoptr(GVariant) commit_data = NULL;
+          g_autoptr(GVariant) commit_metadata = NULL;
+
+          /* We already have the latest commit, resolved it from
+           * the local commit and remove from all results. This way we
+           * avoid pulling this ref from all remotes. */
+
+          if (!ostree_repo_load_commit (child_repo, data->local_commit, &commit_data, NULL, NULL))
+            return FALSE;
+
+          resolve->resolved_commit = g_strdup (data->local_commit);
+          commit_metadata = g_variant_get_child_value (commit_data, 0);
+          g_variant_lookup (commit_metadata, "xa.metadata", "s", &resolve->resolved_metadata);
+
+          remove_ref_from_p2p_results (results, &data->collection_ref);
+        }
+    }
+
+  g_variant_builder_init (&pull_builder, G_VARIANT_TYPE ("a{sv}"));
+  g_variant_builder_add (&pull_builder, "{s@v}", "flags",
+                         g_variant_new_variant (g_variant_new_int32 (OSTREE_REPO_PULL_FLAGS_COMMIT_ONLY)));
+  g_variant_builder_add (&pull_builder, "{s@v}", "inherit-transaction",
+                         g_variant_new_variant (g_variant_new_boolean (TRUE)));
+  pull_options = g_variant_ref_sink (g_variant_builder_end (&pull_builder));
+
+  transaction = flatpak_repo_transaction_start (child_repo, cancellable, error);
+  if (transaction == NULL)
+    return FALSE;
+
+  ostree_repo_pull_from_remotes_async (child_repo, (const OstreeRepoFinderResult * const *)results,
+                                       pull_options, NULL,
+                                       NULL, async_result_cb,
+                                       &pull_result);
+
+  while (pull_result == NULL)
+      g_main_context_iteration (main_context, TRUE);
+
+  if (!ostree_repo_pull_from_remotes_finish (child_repo, pull_result, error))
+    return FALSE;
+
+  for (i = 0; datas[i] != NULL; i++)
+    {
+      FlatpakDirResolveData *data = datas[i];
+      FlatpakDirResolve *resolve = data->resolve;
+      g_autoptr(GVariant) commit_data = NULL;
+      g_autoptr(GVariant) commit_metadata = NULL;
+      g_autofree char *refspec = NULL;
+
+      if (resolve->resolved_commit != NULL)
+        continue;
+
+      refspec = g_strdup_printf ("%s:%s", resolve->remote, resolve->ref);
+      if (!ostree_repo_resolve_rev (child_repo, refspec, FALSE, &resolve->resolved_commit, error))
+        return FALSE;
+
+      if (!ostree_repo_load_commit (child_repo, resolve->resolved_commit, &commit_data, NULL, error))
+        return FALSE;
+
+      commit_metadata = g_variant_get_child_value (commit_data, 0);
+      g_variant_lookup (commit_metadata, "xa.metadata", "s", &resolve->resolved_metadata);
+    }
+
+  /* This will exit and abort the transaction */
+
+  return TRUE;
+
+}
+
+gboolean
+flatpak_dir_resolve_p2p_refs (FlatpakDir               *self,
```

Comment:
Maybe it would be good to document the requirement that each `FlatpakDirResolve` that's passed in has to have `remote` and `ref` non-NULL, and that the corresponding remote has a collection ID? And maybe assert those things as well?